wiki/Client/Release_stream/Lazer/File_storage/en.md

```diff
@@ -0,0 +1,47 @@
+- `Android/data/sh.ppy.osulazer/files/files` (Android).
+
+On desktop platforms, the file store can also be moved in its entirety to another location, using the `Change folder location...` button in client settings.
+
+## Storage structure
+
+osu!(stable)'s storage structure, where data was directly accessible to users as standard files, resulted in a lot of teething issues due to having to explicitly handle unwanted modifications of key files. To avoid this, osu!(lazer) employs a more stringent method of file storage. As such, there is no equivalent of the `Songs` and `Skins` folders. All files that are imported to lazer are stored under filenames that reflect their [SHA-256 hashes](https://en.wikipedia.org/wiki/SHA-2). Mappings to these files are held inside a client database.
+
+For example, a file with the SHA-256 hash of
+
+```
+1a47929b6056d34d25a95eeb2012395ceed66af6f40cc37c898a08482d6325d2
+```
+
+shall be stored under the path of
+
+```
+files/1/1a/1a47929b6056d34d25a95eeb2012395ceed66af6f40cc37c898a08482d6325d2
+```
+
+This allows for saving storage space by preventing duplicates of the same file from being persisted to disk, and prevents users (or other applications) from easily tampering with files that should not be tampered with. All inconveniences compared to stable that arise from this choice have been addressed by either allowing exports or adding new in-game management features, or will be addressed in due time by new features.
+
+At the time of writing, the correct procedure to manually modify skins or beatmaps outside of the facilities available in-game is to export the item, update it, and then reimport it again with the desired changes.
+
+## Migration from osu!(stable)
+
+### Via hard links
+
+On most Windows systems, osu!(lazer) will be able to import data from the stable version of the game without having to create a second copy of the data on disk. This is possible thanks to an operating system feature called *hard links*.
+
+A hard link is conceptually similar to a *shortcut* in that it is a method that allows a user to have the same file available from multiple different places on their filesystem. However, while shortcuts are just plain files that point to a different file (and therefore require additional space), hard links work one level deeper, at the level of the filesystem itself.
+
+In simple terms, two hard-linked files are just two different names for the same piece of physical space on the hard drive. The same data is available through either of those names. This means that once the migration completes using hard links, both the `Songs` folder in stable, and the `files` folder in lazer will contain files which will point to shared underlying data on the disk. Deleting files from one installation will not affect the other. Updating or editing files in one installation will also not affect the other if osu! is used to make changes, rather than external applications.
```

Comment:
i think this will get the point across a bit more concisely

```suggestion
In simple terms, two hard-linked files are just two different names for the same piece of physical space on the hard drive. This means that once the migration completes using hard links, both the `Songs` folder in stable, and the `files` folder in lazer will contain files which will point to shared underlying data on the disk. Deleting files from one installation will not affect the other. Updating or editing files in one installation will also not affect the other, as long as osu! is used to make the changes.
```

you say "Updating or editing files in one installation will also not affect the other" but that kind of implies that changes to a beatmap in stable will not be reflected in lazer or vice versa

could reword to

```suggestion
In simple terms, two hard-linked files are just two different names for the same piece of physical space on the hard drive. This means that once the migration completes using hard links, both the `Songs` folder in stable, and the `files` folder in lazer will contain files which will point to shared underlying data on the disk. Deleting files from one installation will not affect the other. Updating or editing files in one installation will seamlessly be reflected on the other, as long as osu! is used to make the changes.
```

Reply:
> you say "Updating or editing files in one installation will also not affect the other" but that kind of implies that changes to a beatmap in stable will not be reflected in lazer or vice versa

it doesn't, so i'm not sure what point is being made here - the second suggestion seems to just be wrong

Reply:
if the second suggestion is just factually wrong then that's that, seems like i had a misunderstanding. just consider the first one.

but there seems to be a bit of detail glossed over then, because if the data the files point to is the same, you'd intuitively expect changes on one end be reflected on the other, no? i.e. a beatmap imported from stable using hard links would make lazer's files... point to the same data as the beatmap files on stable. idk but if editing one really doesn't affect the other (as if they're using separate data) then that is just contradictory and leaves move questions

Reply:
any change made using either osu! client is deleting one of the hard links and creating a file anew. which is why the other client will not be affected. this is intentional because if you think about how lazer's file store works (each file is stored under its own SHA) then having a change come in from stable *without* breaking the hard link will corrupt lazer's file storage structure.

i'm not sure how to explain this without creating complications in understanding. i'm not sure too many users who aren't technical will care?

Reply:
thank you, that makes sense. i don't think conveying this should be too hard

```suggestion
In simple terms, two hard-linked files are just two different names for the same piece of physical space on the hard drive. This means that once the migration completes using hard links, both the `Songs` folder in stable, and the `files` folder in lazer will contain files which will point to shared underlying data on the disk. Deleting files from one installation will not affect the other. Updating or editing files causes the files to become separate, and will therefore also not affect the other installation, as long as osu! is used to make the changes.
```

maybe i could've come to this perhaps simple conclusion myself if i thought about it harder, but i still think it would be better to spell it out